### ISSUE TITLE
accept_hdr_async : cannot perform an async I/O in the callback because callback does not return a Future

### ISSUE THREAD

**Issue by mahdi-shojaee** [sdroege/async-tungstenite#70]
Hello, thanks for this nice crate. Sometimes it requires to accept only handshakes that pass an **async I/O** validation based on the values from request headers or URI. But currently, we cannot do such operations in the callback of the `accept_hdr_async` function.
Any suggestion?

**Comment by sdroege** [comment]
Can you provide a (non-working) testcase for this, that would make it a bit clearer what you're trying to do. I suspect that it will need a little bit of API to be added, but nothing complicated.

**Comment by mahdi-shojaee** [comment]
Sure:

```rust
async fn ws_accept(
    stream: TcpStream,
    db: <Some DB instance>
) -> std::result::Result<WebSocketStream<TcpStream>, async_tungstenite::tungstenite::Error> {
    async_tungstenite::accept_hdr_async(stream, |req: &Request, res: Response| async {
        if db.validate(req.uri()).await {
            return Ok(res);
        }
        Err(<Some Error>)
    }).await
}
```

**Comment by sdroege** [comment]
I see, so yes that would require adding a bit of API indeed and it's not going to be too easy unfortunately. The callback goes down deep into tungstenite [here](https://docs.rs/tungstenite/0.11.1/src/tungstenite/handshake/server.rs.html#244). That code knows nothing about async.

It would be necessary to extend the callback there with a way to return an `std::io::Error` so we could return `WouldBlock` here from our side to make it work with futures, and call into the handshake again once the future is ready again to be polled.

**Comment by mahdi-shojaee** [comment]
Yes, I see, I think adding this will be so useful because it will let us dynamically accept connections just if for example some remote service validates the request. Any plan to add this?

**Comment by mahdi-shojaee** [comment]
Because doing this synchronously is horrible.

**Comment by sdroege** [comment]
Yeah it seems useful to have, I agree :) I don't think I will have time for working on that anytime soon, but I'd be happy to review/help anybody who would like to implement it.

**Comment by databasedav** [comment]
i'm a rust beginner but i'd like to help implement this if i can, what exactly is required? being able to do async io in the connect callback is crucial :)

**Comment by sdroege** [comment]
@databasedav https://github.com/sdroege/async-tungstenite/issues/70#issuecomment-748497017 has the basic idea how this could be implemented

**Comment by databasedav** [comment]
@sdroege can u expand on this implementation? i'm thinking creating an async version of the `Callback` trait `AsyncCallback`

```rust
pub trait AsyncCallback: Sized {
    async fn on_request(
        self,
        request: &Request,
        response: Response,
    ) -> StdResult<Response, ErrorResponse>;
}
```
(which we would need the `async-trait` crate to make work) and then making async versions of the methods that wrap the `on_request`. is this the way to go? i might be ignoring some rust limitations, this is the sort of path i've taken when migrating similar python
